### PR TITLE
Create outbound message store

### PR DESCRIPTION
**Pull request by rudigiesler**
Create an outbound message store that stores the link between the unique message id and the event URL.

### CODE REVIEW

**Comment by rudigiesler** [comment]
Ready for review.

File: junebug/stores.py [hunk 1]
```diff
@@ -1,15 +1,16 @@
-from twisted.internet.defer import inlineCallbacks, returnValue
+from twisted.internet.defer import inlineCallbacks, returnValue, succeed
 from vumi.message import TransportUserMessage
 
 
 class BaseStore(object):
+    KEYSPACE = 'basestore'
```
Comment (hodgestar) [inline junebug/stores.py]: This should go below the docstring, otherwise the docstring is not a docstring. :)

File: junebug/stores.py [hunk 2]
```diff
@@ -21,6 +22,11 @@ def _redis_op(self, func, id, *args, **kwargs):
         yield self.redis.expire(id, self.ttl)
         returnValue(val)
 
+    def get_key(self, channel_id, message_id):
+        '''Gets the key where a message would be stored, using the channel and
+        message ids'''
+        return '%s:%s:%s' % (channel_id, self.KEYSPACE, message_id)
```
Comment (hodgestar) [inline junebug/stores.py]: I'm wondering if `get_key` should just be implemented by sub-classes? The key returned seems use-case specific.

File: junebug/stores.py [hunk 3]
```diff
@@ -41,21 +47,37 @@ def load_property(self, id, key):
...
         if msg_json is None:
             returnValue(None)
         returnValue(TransportUserMessage.from_json(msg_json))
+
+
+class OutboundMessageStore(BaseStore):
+    KEYSPACE = 'outbound_messages'
+    '''Stores the event url, in order to look it up when deciding where events
+    should go'''
+    def store_vumi_message(self, channel_id, api_request, message):
```
Comment (hodgestar) [inline junebug/stores.py]: Should be named `store_event_url`?

Comment (hodgestar) [inline junebug/stores.py]: Would be happier if this just took just the `event_url` rather than the `api_request`.

File: junebug/stores.py (hunk 2 above)
Comment (hodgestar) [inline junebug/stores.py]: I'm kind of thinking we want to prefix this key with `channels:` in case we need to store things that are not related to a particular channel at some point.

**Comment by hodgestar** [comment]
Left some comments but looks good.

File: junebug/stores.py (hunk 2 above)
Comment (rudigiesler) [inline junebug/stores.py]: I did it this way to keep consistency. Currently we use `<channel-id>:properties` to store the channel properties. Should we be changing that to `channels:<channel-id>:properties`?

File: junebug/stores.py (hunk 3 above)
Comment (rudigiesler) [inline junebug/stores.py]: The idea I had doing it this way is that we might want to store other things in the future, and then we wouldn't need to change all the places that this is called, because all the information is here already (say if we decided we also wanted to store the content). Then we would just change the call to a `store_all` with a dictionary, and we keep it to only one redis call.

We could change it to just store the event url, if you think that's a better solution for now.

File: junebug/stores.py (hunk 2 above)
Comment (rudigiesler) [inline junebug/stores.py]: The reason I changed to putting it by the base class is that I would be repeating essentially the same thing in all the subclasses, and I wanted to reduce that repetition. Maybe the base class should have a method that takes in n strings, and returns a key, which the subclasses can use in their own key making functions.

File: junebug/stores.py [hunk 4]
```diff
@@ -43,19 +50,37 @@ def load_property(self, id, key):
...
+    should go'''
+    KEYSPACE = 'outbound_messages'
+
+    def store_vumi_message(self, channel_id, api_request, message):
+        '''Stores the event_url'''
+        if api_request.get('event_url') is not None:
+            key = self.get_key(channel_id, message.get('message_id'))
+            return self.store_property(
+                key, 'event_url', api_request['event_url'])
+        return succeed(None)
```
Comment (justinvdm) [inline junebug/stores.py]: Wondering whether stores should know about api requests and how they look. Would it maybe be better to pass in `event_url` directly?

File: junebug/stores.py [hunk 5]
```diff
@@ -43,19 +50,37 @@ def load_property(self, id, key):
...
         returnValue(TransportUserMessage.from_json(msg_json))
+
+
+class OutboundMessageStore(BaseStore):
+    '''Stores the event url, in order to look it up when deciding where events
+    should go'''
+    KEYSPACE = 'outbound_messages'
+
+    def store_vumi_message(self, channel_id, api_request, message):
+        '''Stores the event_url'''
```
Comment (justinvdm) [inline junebug/stores.py]: If the plan is for this method to only handle storing event urls, maybe the name should suggest that it stores event urls associated to a message?

Comment (justinvdm) [inline junebug/stores.py]: Ah, sorry, just saw [this discussion](https://github.com/praekelt/junebug/pull/29#discussion_r39950445)

**Comment by justinvdm** [comment]
Left minor comment, looks good :)

File: junebug/stores.py (hunk 2 above)
Comment (hodgestar) [inline junebug/stores.py]: Sounds good.

**Comment by rudigiesler** [comment]
Ready for another review

**Comment by justinvdm** [comment]
:+1:

**Comment by rudigiesler** [comment]
Last commit changed the parameter order. Ready for review again.

**Comment by hodgestar** [comment]
:+1:
